sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/planning/SelectedFieldSuite.scala

```diff
@@ -0,0 +1,387 @@
+  }
+
+  testSelect("col7.field1", "col7[0].field1 as foo", "col7.field1[0] as foo") {
+    StructField("col7", ArrayType(StructType(
+      StructField("field1", IntegerType, nullable = false) :: Nil)))
+  }
+
+  testSelect("col7.field2.subfield1") {
+    StructField("col7", ArrayType(StructType(
+      StructField("field2", StructType(
+        StructField("subfield1", IntegerType, nullable = false) :: Nil)) :: Nil)))
+  }
+
+  testSelect("col7.field3.subfield1") {
+    StructField("col7", ArrayType(StructType(
+      StructField("field3", ArrayType(StructType(
+        StructField("subfield1", IntegerType, nullable = false) :: Nil))) :: Nil)))
+  }
+
+  testSelect("col8.field1", "col8[0].field1 as foo", "col8.field1[0] as foo", "col8[0].field1[0] as foo") {
+    StructField("col8", ArrayType(StructType(
+      StructField("field1", ArrayType(IntegerType, containsNull = false), nullable = false) :: Nil)))
+  }
+
+  def assertResult(expected: StructField)(actual: StructField)(selectExpr: String): Unit = {
+    try {
+      super.assertResult(expected)(actual)
+    } catch {
+      case ex: TestFailedException =>
+        // Print some helpful diagnostics in the case of failure
+        alert("Expected SELECT \"" + selectExpr + "\" to select the schema\n" +
+          indent(StructType(expected :: Nil).treeString) +
+          indent("but it actually selected\n") +
+          indent(StructType(actual :: Nil).treeString) +
+          indent("Note that expected.dataType.sameType(actual.dataType) = " +
+          expected.dataType.sameType(actual.dataType)))
+        throw ex
+    }
+  }
+
+  // Test that the given SELECT expressions prune the test schema to the single-column schema
+  // defined by the given field
+  private def testSelect(selectExpr: String, otherSelectExprs: String*)(expected: StructField) {
+    val selectExprs = selectExpr +: otherSelectExprs
+    test(s"SELECT " + selectExprs.map(s => s""""$s"""").mkString(", ") + " should select the schema\n" +
+        indent(StructType(expected :: Nil).treeString)) {
+      for (selectExpr <- selectExprs) {
+        assertSelect(selectExpr, expected)
+      }
+    }
+  }
+
+  private def assertSelect(expr: String, expected: StructField) = {
+    unapplySelect(expr) match {
+      case Some(field) =>
+        assertResult(expected)(field)(expr)
+      case None =>
+        val failureMessage =
+          "Failed to select a field from " + expr + ". " +
+          "Expected:\n" +
+          StructType(expected :: Nil).treeString
+        fail(failureMessage)
+    }
+  }
+
+  private def unapplySelect(expr: String) = {
+    val parsedExpr =
+      CatalystSqlParser.parseExpression(expr) match {
+        case namedExpr: NamedExpression => namedExpr
+      }
+    val select = testRelation.select(parsedExpr)
+    val analyzed = select.analyze
+    SelectedField.unapply(analyzed.expressions.head)
+  }
+
+  private def indent(string: String) = string.replaceAll("(?m)^", "   ")
```

Comment:
Is it for add 4 spaces? I think we can remove this and just use 4 spaces. Virtually the code lengths are same.

Reply:
This method indents every line in a string by four spaces. For example, https://regex101.com/r/teuQcA/1.

I'll add an explanatory comment in my next push.